Automattic/woocommerce-subscriptions-core
# Prevent a post record re-sync on every subscription get

Fixes #268 

## Description

On HPOS sites with syncing enabled, every time `wcs_get_subscription()` is called, WC will incorrectly determine that the subscription needs to be resynced when nothing was changed. 

The reason this erroneous resync occurs this is because WC determines if syncing needs to occur in the `read_multiple()` function before we've had a chance to set our subscription specific properties. 

So the comparison WC does between the HPOS and Post subscription objects aren't the same - the HPOS one is incomplete, the POST is complete.

This PR fixes that by utilising the changes I made to WC core in https://github.com/woocommerce/woocommerce/pull/35829. This change allows us to set Subscription properties in before WC do the compare, meaning the "Do we need to resync" check runs on completely loaded objects.

## How to test this PR

As far as I could tell, this resync wasn't causing and data corruption but given this change requires changes to a fundamental part of our datastore and changes to WC core, it's important that we patch this relatively early. 

1. Install the latest WC nightly build.
1. Apply the following settings see below.
    - HPOS tables enabled.
    - Syncing enabled. 
1. Trigger any `wcs_get_subscription()` call.
      - either view the my account view subscription page
      - using wp console call `wcs_get_subscription()`.
1. Inside the `maybe_sync_order()` function add a breakpoint, error log or add something to let you know the `migrate_post_record()` function is being called.
     - On `trunk` the migrate_post_record() function is called everytime. 
     - On this branch the migrate_post_record function should only be called when something is different. You can experiment with this by manually changing a subscription property in the HPOS or WP synced post tables. 


> **Warning**
> Syncing from the post record to the HPOS record doesn't currently work for subscription dates (eg next payment date). We will need to fix this separately. 
> - https://github.com/Automattic/woocommerce-subscriptions-core/issues/355

![image](https://user-images.githubusercontent.com/8490476/209262441-ca168050-700f-4a9a-bb0d-369845fcb342.png)


## Product impact
